src/tactic/ring2.lean

```diff
@@ -2,90 +2,104 @@
 Copyright (c) 2018 Mario Carneiro. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Mario Carneiro
+-/
+
+import tactic.ring data.num.lemmas data.tree
+import tactic.converter.interactive
+
+/-!
+# ring2
 
 An experimental variant on the `ring` tactic that uses computational
 reflection instead of proof generation. Useful for kernel benchmarking.
 -/
-import tactic.ring data.num.lemmas
-import tactic.converter.interactive
-
-namespace tactic.ring2
 
-@[derive has_reflect]
-inductive {u} tree (α : Type u) : Type u
-| nil {} : tree
-| node : α → tree → tree → tree
-
-def tree.get {α} [has_zero α] : pos_num → tree α → α
-| _                tree.nil            := 0
-| pos_num.one      (tree.node a t₁ t₂) := a
-| (pos_num.bit0 n) (tree.node a t₁ t₂) := t₁.get n
-| (pos_num.bit1 n) (tree.node a t₁ t₂) := t₂.get n
-
-def tree.of_rbnode {α} : rbnode α → tree α
-| rbnode.leaf               := tree.nil
-| (rbnode.red_node l a r)   :=
-  tree.node a (tree.of_rbnode l) (tree.of_rbnode r)
-| (rbnode.black_node l a r) :=
-  tree.node a (tree.of_rbnode l) (tree.of_rbnode r)
-
-def tree.index_of {α} (lt : α → α → Prop) [decidable_rel lt]
-  (x : α) : tree α → option pos_num
-| tree.nil := none
-| (tree.node a t₁ t₂) :=
-  match cmp_using lt x a with
-  | ordering.lt := pos_num.bit0 <$> tree.index_of t₁
-  | ordering.eq := some pos_num.one
-  | ordering.gt := pos_num.bit1 <$> tree.index_of t₂
-  end
+namespace tree
 
-meta def tree.reflect' (u : level) (α : expr) : tree expr → expr
+/-- `(reflect' t u α)` turns a tree `(t: tree expr)` of reflected values
+of type `α` at level `u` into an `expr` which concretizes to a `tree α`
+containing the concretizations of the `expr`s from the original `t`. -/
```

Comment:
Hm, this comment reads a bit convoluted. The way I would explain this is that we can quasiquote a `tree expr`, with unquotations at the leaves, where the leaves represent elements of type alpha (at level u).

Reply:
I am also not too happy with this comment. The issue here is that the terminology around quoting is varied and inconsistent. I've heard the terms reflection, reification, quotation and concretization all used in this context and they always confuse me. Perhaps "quoting" is closest to existing Lean terminology, so I will use that instead.

Reply:
I changed it to use "(quasi)quotation"/"reification" but avoided the use of "unquotation". The reason is that a quick search for "unquotation" comes up with nothing, while the former terms have some explanations on the Haskell wiki.

Reply:
(One nitpick, this tree type has values at every node rather than just the leaves).